File: js/src/absolute-path.js
// @flow

// Copyright 2016 Attic Labs, Inc. All rights reserved.
// Licensed under the Apache License, version 2.0:
// http://www.apache.org/licenses/LICENSE-2.0

import {invariant} from './assert.js';
import {datasetRe} from './dataset.js';
import Database from './database.js';
import Hash, {stringLength} from './hash.js';
import Path from './path.js';
import type Value from './value.js';

const datasetCapturePrefixRe = new RegExp('^(' + datasetRe.source + ')');

/**
 * An AbsolutePath is a Path relative to either a dataset head, or a hash.
 *
 * E.g. in a spec like `http://demo.noms.io::foo.bar` this is the `foo.bar` component, or in
 * `http://demo.noms.io::#abcdef.bar` the `#abcdef.bar` component.
 *
 * See https://github.com/attic-labs/noms/blob/master/doc/spelling.md.
 */
export default class AbsolutePath {
  /** The dataset ID that `path` is in, or `''` if none. */
  dataset: string;

  /** The hash the that `path` is in, if any. */
  hash: Hash | null;

  /** Path relative to either `dataset` or `hash`. */
  path: Path;

  /**
   * Returns `str` parsed as an AbsolutePath. Throws a `SyntaxError` if `str` isn't a valid path.
   */
  static parse(str: string): AbsolutePath {
    if (str === '') {
      throw new SyntaxError('Empty path');
    }

    let dataset = '';
    let hash = null;
    let pathStr = '';

    if (str[0] === '#') {
      const tail = str.slice(1);
      if (tail.length < stringLength) {
        throw new SyntaxError(`Invalid hash: ${tail}`);
      }

      const hashStr = tail.slice(0, stringLength);
      hash = Hash.parse(hashStr);
      if (hash === null) {
        throw new SyntaxError(`Invalid hash: ${hashStr}`);
      }

      pathStr = tail.slice(stringLength);
    } else {
      const parts = datasetCapturePrefixRe.exec(str);
      if (!parts) {
        throw new SyntaxError(`Invalid dataset name: ${str}`);
      }

      invariant(parts.length === 2);
      dataset = parts[1];
      pathStr = str.slice(parts[0].length);
    }

    if (pathStr.length === 0) {
      return new AbsolutePath(dataset, hash, new Path());
    }

    const path = Path.parse(pathStr);
    return new AbsolutePath(dataset, hash, path);
  }

  constructor(dataset: string, hash: Hash | null, path: Path) {
    this.dataset = dataset;
    this.hash = hash;
    this.path = path;
  }

  async resolve(db: Database): Promise<Value | null> {
    let val = null;
    if (this.dataset !== '') {
      val = await db.head(this.dataset);
    } else if (this.hash !== null) {
      val = await db.readValue(this.hash);
    } else {
      throw new Error('unreachable');
    }

    if (val === undefined) {
      val = null;
    }
    return val === null ? null : this.path.resolve(val);
  }

  toString(): string {
    if (this.dataset !== '') {
      return this.dataset + this.path.toString();
    }
    if (this.hash !== null) {
      return '#' + this.hash.toString() + this.path.toString();
    }
    throw new Error('unreachable');
  }
}

File: go/store/spec/absolute_path.go
// Copyright 2019 Dolthub, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// This file incorporates work covered by the following copyright and
// permission notice:
//
// Copyright 2016 Attic Labs, Inc. All rights reserved.
// Licensed under the Apache License, version 2.0:
// http://www.apache.org/licenses/LICENSE-2.0

package spec

import (
	"context"
	"errors"

	"github.com/dolthub/dolt/go/store/datas"
	"github.com/dolthub/dolt/go/store/hash"
	"github.com/dolthub/dolt/go/store/types"
)

// AbsolutePath describes the location of a Value within a Noms database.
//
// To locate a value relative to some other value, see Path. To locate a value
// globally, see Spec.
//
// For more on paths, absolute paths, and specs, see:
// https://github.com/attic-labs/noms/blob/master/doc/spelling.md.
type AbsolutePath struct {
	// Dataset is the dataset this AbsolutePath is rooted at. Only one of
	// Dataset and Hash should be set.
	Dataset string
	// Hash is the hash this AbsolutePath is rooted at. Only one of Dataset and
	// Hash should be set.
	Hash hash.Hash
}

// NewAbsolutePath attempts to parse 'str' and return an AbsolutePath.
func NewAbsolutePath(str string) (AbsolutePath, error) {
	if len(str) == 0 {
		return AbsolutePath{}, errors.New("empty path")
	}

	var h hash.Hash
	var dataset string

	if str[0] == '#' {
		tail := str[1:]
		if len(tail) < hash.StringLen {
			return AbsolutePath{}, errors.New("invalid hash: " + tail)
		}

		hashStr := tail[:hash.StringLen]
		if h2, ok := hash.MaybeParse(hashStr); ok {
			h = h2
		} else {
			return AbsolutePath{}, errors.New("invalid hash: " + hashStr)
		}
	} else {
		// This form is only used in the noms command. Commands like `noms show` that use a path with '.' separation will
		// no longer work
		dataset = str
	}

	return AbsolutePath{Hash: h, Dataset: dataset}, nil
}

// Resolve returns the Value reachable by 'p' in 'db'.
func (p AbsolutePath) Resolve(ctx context.Context, db datas.Database, vrw types.ValueReadWriter) (val types.Value, err error) {
	if len(p.Dataset) > 0 {
		var ok bool
		ds, err := db.GetDataset(ctx, p.Dataset)
		if err != nil {
			return nil, err
		}

		if val, ok = ds.MaybeHead(); !ok {
			val = nil
		}
	} else if !p.Hash.IsEmpty() {
		var err error
		val, err = vrw.ReadValue(ctx, p.Hash)
		if err != nil {
			return nil, err
		}
	} else {
		panic("Unreachable")
	}

	return
}

func (p AbsolutePath) IsEmpty() bool {
	return p.Dataset == "" && p.Hash.IsEmpty()
}

func (p AbsolutePath) String() (str string) {
	if p.IsEmpty() {
		return ""
	}

	if len(p.Dataset) > 0 {
		str = p.Dataset
	} else if !p.Hash.IsEmpty() {
		str = "#" + p.Hash.String()
	} else {
		panic("Unreachable")
	}

	return str
}
